File: packages/coder_agent/lib/src/gitignore.dart
import 'dart:io' show FileSystemException;

import 'package:file/file.dart' as file_api;

/// One parsed line of a gitignore file.
///
/// A pattern is compiled to a [RegExp] once, because the walker tests every
/// pattern against every entry it visits and a repository has far more entries
/// than patterns.
class GitignorePattern {
  /// Creates a [GitignorePattern].
  const GitignorePattern({
    required this.negated,
    required this.directoryOnly,
    required this.expression,
  });

  /// Parses one line, or returns null when the line carries no pattern.
  ///
  /// Blank lines and comments are the two cases git ignores outright.
  static GitignorePattern? parse(String line) {
    var body = _stripTrailingSpaces(line);
    if (body.isEmpty) return null;
    // A comment cannot be escaped away by anything but a leading backslash.
    if (body.startsWith('#')) return null;
    if (body.startsWith(r'\#')) body = body.substring(1);

    var negated = false;
    if (body.startsWith('!')) {
      negated = true;
      body = body.substring(1);
    } else if (body.startsWith(r'\!')) {
      body = body.substring(1);
    }
    if (body.isEmpty) return null;

    var directoryOnly = false;
    if (body.endsWith('/') && !body.endsWith(r'\/')) {
      directoryOnly = true;
      body = body.substring(0, body.length - 1);
    }
    if (body.isEmpty) return null;

    // A slash anywhere but the end anchors the pattern to the directory the
    // gitignore file lives in; without one it matches at any depth.
    final anchored = body.substring(0, body.length - 1).contains('/');
    if (body.startsWith('/')) body = body.substring(1);
    if (body.isEmpty) return null;

    return GitignorePattern(
      negated: negated,
      directoryOnly: directoryOnly,
      expression: RegExp('^${_toRegExp(body, anchored: anchored)}\$'),
    );
  }

  /// Whether a match re-includes the entry instead of ignoring it.
  final bool negated;

  /// Whether the pattern only ever matches directories.
  final bool directoryOnly;

  /// Compiled matcher, applied to a path relative to the owning directory.
  final RegExp expression;

  /// Whether [relativePath] matches, given whether it names a directory.
  bool matches(String relativePath, {required bool isDirectory}) {
    if (directoryOnly && !isDirectory) return false;
    return expression.hasMatch(relativePath);
  }

  /// Drops trailing whitespace unless the last space was escaped.
  static String _stripTrailingSpaces(String line) {
    var end = line.length;
    while (end > 0 && (line[end - 1] == ' ' || line[end - 1] == '\t')) {
      // An escaped space is part of the pattern, and the backslash before it
      // only counts when it is not itself escaped.
      var backslashes = 0;
      var index = end - 2;
      while (index >= 0 && line[index] == r'\') {
        backslashes += 1;
        index -= 1;
      }
      if (backslashes.isOdd) break;
      end -= 1;
    }
    return line.substring(0, end);
  }

  /// Translates gitignore glob syntax into a regular expression body.
  static String _toRegExp(String pattern, {required bool anchored}) {
    final buffer = StringBuffer();
    // An unanchored pattern may start at any depth, which git expresses as an
    // implicit leading `**/`.
    if (!anchored) buffer.write('(?:.*/)?');
    var index = 0;
    while (index < pattern.length) {
      final char = pattern[index];
      if (char == r'\') {
        // The escaped character is a literal, whatever it is.
        if (index + 1 < pattern.length) {
          buffer.write(RegExp.escape(pattern[index + 1]));
          index += 2;
        } else {
          buffer.write(RegExp.escape(char));
          index += 1;
        }
        continue;
      }
      if (char == '*') {
        final isDouble =
            index + 1 < pattern.length && pattern[index + 1] == '*';
        if (!isDouble) {
          // A single star never crosses a directory boundary.
          buffer.write('[^/]*');
          index += 1;
          continue;
        }
        final followedBySlash =
            index + 2 < pattern.length && pattern[index + 2] == '/';
        final atEnd = index + 2 >= pattern.length;
        if (atEnd) {
          // A trailing `**` matches everything below, including nothing.
          buffer.write('.*');
          index += 2;
          continue;
        }
        if (followedBySlash) {
          // `**/` collapses to "zero or more directories", so `a/**/b` also
          // matches `a/b`.
          buffer.write('(?:.*/)?');
          index += 3;
          continue;
        }
        // `**` not delimited by slashes is treated as a single star by git.
        buffer.write('[^/]*');
        index += 2;
        continue;
      }
      if (char == '?') {
        buffer.write('[^/]');
        index += 1;
        continue;
      }
      if (char == '[') {
        final closing = _findClassEnd(pattern, index);
        if (closing == -1) {
          buffer.write(RegExp.escape(char));
          index += 1;
          continue;
        }
        buffer.write(_toCharacterClass(pattern.substring(index, closing + 1)));
        index = closing + 1;
        continue;
      }
      buffer.write(RegExp.escape(char));
      index += 1;
    }
    return buffer.toString();
  }

  /// Index of the `]` closing the class opened at [start], or -1.
  static int _findClassEnd(String pattern, int start) {
    var index = start + 1;
    if (index < pattern.length &&
        (pattern[index] == '!' || pattern[index] == '^')) {
      index += 1;
    }
    // A `]` in the first position is a literal member, not the terminator.
    if (index < pattern.length && pattern[index] == ']') index += 1;
    while (index < pattern.length) {
      if (pattern[index] == r'\') {
        index += 2;
        continue;
      }
      if (pattern[index] == ']') return index;
      index += 1;
    }
    return -1;
  }

  /// Rewrites a glob character class as a regular-expression one.
  static String _toCharacterClass(String glob) {
    final inner = glob.substring(1, glob.length - 1);
    // Glob negates with `!`; a regular expression uses `^`.
    final negatedClass = inner.startsWith('!') || inner.startsWith('^');
    final body = negatedClass ? inner.substring(1) : inner;
    // A class must never match a separator, or `[a-z]` would cross directories.
    return '[${negatedClass ? '^/' : ''}${body.replaceAll(r'\', r'\\')}]';
  }
}

/// The patterns of one gitignore file, and the directory they apply to.
class GitignoreSource {
  /// Creates a [GitignoreSource].
  const GitignoreSource({required this.basePath, required this.patterns});

  /// Parses [contents] as a gitignore file governing [basePath].
  ///
  /// [basePath] is relative to the workspace root and uses `/` separators; the
  /// root itself is the empty string.
  factory GitignoreSource.parse(String basePath, String contents) =>
      GitignoreSource(
        basePath: basePath,
        patterns: contents
            .split('\n')
            .map((line) => GitignorePattern.parse(_stripCarriageReturn(line)))
            .whereType<GitignorePattern>()
            .toList(growable: false),
      );

  static String _stripCarriageReturn(String line) =>
      line.endsWith('\r') ? line.substring(0, line.length - 1) : line;

  /// Workspace-relative directory this file governs.
  final String basePath;

  /// Patterns in file order; the last one that matches decides.
  final List<GitignorePattern> patterns;

  /// Whether this file ignores [relativePath], or null when it says nothing.
  ///
  /// [relativePath] is relative to the workspace root. Null is distinct from
  /// false: it lets a lower-precedence source keep its own verdict.
  bool? verdict(String relativePath, {required bool isDirectory}) {
    final scoped = _scope(relativePath);
    if (scoped == null) return null;
    bool? decision;
    // Later patterns override earlier ones, so the whole file is scanned rather
    // than short-circuiting on the first hit.
    for (final pattern in patterns) {
      if (pattern.matches(scoped, isDirectory: isDirectory)) {
        decision = !pattern.negated;
      }
    }
    return decision;
  }

  /// Re-expresses [relativePath] relative to [basePath], or null if outside.
  String? _scope(String relativePath) {
    if (basePath.isEmpty) return relativePath;
    final prefix = '$basePath/';
    if (!relativePath.startsWith(prefix)) return null;
    return relativePath.substring(prefix.length);
  }
}

/// Decides whether git would ignore a workspace path.
///
/// Sources are held lowest-precedence first, which is the order git applies
/// them: global excludes, then `.git/info/exclude`, then every `.gitignore`
/// from the workspace root down to the entry's own directory.
class GitignoreMatcher {
  /// Creates a [GitignoreMatcher] over an ordered list of [sources].
  GitignoreMatcher(this.sources);

  /// A matcher that ignores nothing.
  GitignoreMatcher.empty() : sources = const <GitignoreSource>[];

  /// Sources in ascending precedence.
  final List<GitignoreSource> sources;

  /// Returns a matcher with [source] appended at the highest precedence.
  GitignoreMatcher push(GitignoreSource source) =>
      GitignoreMatcher(<GitignoreSource>[...sources, source]);

  /// Whether git would ignore [relativePath].
  bool isIgnored(String relativePath, {required bool isDirectory}) {
    var ignored = false;
    for (final source in sources) {
      final verdict = source.verdict(relativePath, isDirectory: isDirectory);
      if (verdict != null) ignored = verdict;
    }
    return ignored;
  }
}

/// Where a loader may look for git configuration outside the workspace.
///
/// This is passed in rather than read from the ambient environment so that a
/// test can never pick up the excludes file of whoever is running it — the
/// same reason every other host boundary here is a port.
class GitignoreEnvironment {
  /// Creates a [GitignoreEnvironment].
  const GitignoreEnvironment({this.home, this.xdgConfigHome});

  /// An environment with no user-level git configuration at all.
  const GitignoreEnvironment.none() : home = null, xdgConfigHome = null;

  /// Reads the locations out of [environment], for use in a composition root.
  factory GitignoreEnvironment.fromEnvironment(
    Map<String, String> environment,
  ) {
    String? nonEmpty(String? value) =>
        value != null && value.isNotEmpty ? value : null;
    return GitignoreEnvironment(
      home:
          nonEmpty(environment['HOME']) ?? nonEmpty(environment['USERPROFILE']),
      xdgConfigHome: nonEmpty(environment['XDG_CONFIG_HOME']),
    );
  }

  /// The user's home directory, or null when there is none to consult.
  final String? home;

  /// `XDG_CONFIG_HOME`, which outranks [home] when set.
  final String? xdgConfigHome;
}

/// Loads the gitignore sources that apply outside the working tree.
///
/// These are the two git consults before any `.gitignore`: the user's global
/// excludes file and the repository's own `.git/info/exclude`.
class GitignoreLoader {
  /// Creates a [GitignoreLoader].
  const GitignoreLoader({
    required this._fileSystem,
    this.environment = const GitignoreEnvironment.none(),
  });

  /// Where user-level git configuration lives.
  final GitignoreEnvironment environment;

  final file_api.FileSystem _fileSystem;

  /// Sources that apply to every path in the workspace at [workspaceRoot].
  GitignoreMatcher baseMatcher(String workspaceRoot) {
    final sources = <GitignoreSource>[];
    final global = _globalExcludesFile(workspaceRoot);
    if (global != null) {
      final contents = _read(global);
      if (contents != null) sources.add(GitignoreSource.parse('', contents));
    }
    final info = _read(
      _fileSystem.path.join(workspaceRoot, '.git', 'info', 'exclude'),
    );
    if (info != null) sources.add(GitignoreSource.parse('', info));
    return GitignoreMatcher(sources);
  }

  /// The `.gitignore` governing [directory], or null when there is none.
  ///
  /// [relativeBase] is the workspace-relative, `/`-separated form of
  /// [directory], which is the coordinate space patterns are matched in.
  GitignoreSource? sourceForDirectory(String directory, String relativeBase) {
    final contents = _read(_fileSystem.path.join(directory, '.gitignore'));
    return contents == null
        ? null
        : GitignoreSource.parse(relativeBase, contents);
  }

  /// Resolves `core.excludesFile`, falling back to git's documented default.
  String? _globalExcludesFile(String workspaceRoot) {
    final configured = _configuredExcludesFile(workspaceRoot);
    if (configured != null) return _expandHome(configured);
    final xdg = environment.xdgConfigHome;
    if (xdg != null) return _fileSystem.path.join(xdg, 'git', 'ignore');
    final home = environment.home;
    return home == null
        ? null
        : _fileSystem.path.join(home, '.config', 'git', 'ignore');
  }

  /// Reads `core.excludesFile` from the config files git would consult.
  ///
  /// Later files win, matching git's precedence: system and user config are
  /// overridden by the repository's own.
  String? _configuredExcludesFile(String workspaceRoot) {
    final home = environment.home;
    final xdg = environment.xdgConfigHome;
    final candidates = <String>[
      if (xdg != null) _fileSystem.path.join(xdg, 'git', 'config'),
      if (home != null) _fileSystem.path.join(home, '.gitconfig'),
      _fileSystem.path.join(workspaceRoot, '.git', 'config'),
    ];
    String? found;
    for (final candidate in candidates) {
      final contents = _read(candidate);
      if (contents == null) continue;
      final value = _readCoreExcludesFile(contents);
      if (value != null) found = value;
    }
    return found;
  }

  /// Extracts `excludesfile` from the `[core]` section of a git config.
  ///
  /// Only that one key is understood; anything else in the file is skipped
  /// rather than parsed, because nothing here needs it.
  static String? _readCoreExcludesFile(String contents) {
    var inCore = false;
    String? value;
    for (final raw in contents.split('\n')) {
      final line = raw.trim();
      if (line.isEmpty || line.startsWith('#') || line.startsWith(';')) {
        continue;
      }
      if (line.startsWith('[')) {
        final close = line.indexOf(']');
        final header = close > 1 ? line.substring(1, close).trim() : '';
        inCore = header.toLowerCase() == 'core';
        continue;
      }
      if (!inCore) continue;
      final separator = line.indexOf('=');
      if (separator == -1) continue;
      final key = line.substring(0, separator).trim().toLowerCase();
      if (key != 'excludesfile') continue;
      value = _unquote(line.substring(separator + 1).trim());
    }
    return value;
  }

  static String _unquote(String value) =>
      value.length >= 2 && value.startsWith('"') && value.endsWith('"')
      ? value.substring(1, value.length - 1)
      : value;

  String _expandHome(String path) {
    final home = environment.home;
    if (home == null) return path;
    if (path == '~') return home;
    // Only `~/` is expanded; `~user` names another account, which nothing here
    // can resolve and git itself only supports through the OS.
    return path.startsWith('~/')
        ? _fileSystem.path.join(home, path.substring(2))
        : path;
  }

  String? _read(String path) {
    try {
      final file = _fileSystem.file(path);
      return file.existsSync() ? file.readAsStringSync() : null;
    } on FileSystemException {
      return null;
    }
  }
}

File: packages/daemon/builtin_plugins/tinest.files/lua/gitignore.lua
local gitignore = {}

local function normalize(path)
  local value = tostring(path or ""):gsub("\\\\", "/")
  value = value:gsub("^%./", ""):gsub("/+$", "")
  if value == "." then return "" end
  return value
end

local function strip_trailing_spaces(line)
  local finish = #line
  while finish > 0 do
    local char = line:sub(finish, finish)
    if char ~= " " and char ~= "\t" then break end
    local backslashes = 0
    local cursor = finish - 1
    while cursor > 0 and line:sub(cursor, cursor) == "\\" do
      backslashes = backslashes + 1
      cursor = cursor - 1
    end
    if backslashes % 2 == 1 then break end
    finish = finish - 1
  end
  return line:sub(1, finish)
end

local function class_end(pattern, start)
  local cursor = start + 1
  local first = pattern:sub(cursor, cursor)
  if first == "!" or first == "^" then cursor = cursor + 1 end
  if pattern:sub(cursor, cursor) == "]" then cursor = cursor + 1 end
  while cursor <= #pattern do
    local char = pattern:sub(cursor, cursor)
    if char == "\\" then
      cursor = cursor + 2
    elseif char == "]" then
      return cursor
    else
      cursor = cursor + 1
    end
  end
  return nil
end

local function parse_class(pattern, start, finish)
  local cursor = start + 1
  local first = pattern:sub(cursor, cursor)
  local negated = first == "!" or first == "^"
  if negated then cursor = cursor + 1 end
  local characters = {}
  while cursor < finish do
    local char = pattern:sub(cursor, cursor)
    if char == "\\" and cursor + 1 < finish then
      cursor = cursor + 1
      char = pattern:sub(cursor, cursor)
    end
    characters[#characters + 1] = char
    cursor = cursor + 1
  end
  return {kind = "class", negated = negated, characters = characters}
end

local function tokenize(pattern)
  local tokens = {}
  local cursor = 1
  while cursor <= #pattern do
    local char = pattern:sub(cursor, cursor)
    if char == "\\" then
      if cursor < #pattern then
        cursor = cursor + 1
        char = pattern:sub(cursor, cursor)
      end
      tokens[#tokens + 1] = {kind = "literal", value = char}
      cursor = cursor + 1
    elseif char == "*" then
      local double = pattern:sub(cursor + 1, cursor + 1) == "*"
      if double and cursor + 1 == #pattern then
        tokens[#tokens + 1] = {kind = "double_star_end"}
        cursor = cursor + 2
      elseif double and pattern:sub(cursor + 2, cursor + 2) == "/" then
        tokens[#tokens + 1] = {kind = "double_star_directories"}
        cursor = cursor + 3
      else
        tokens[#tokens + 1] = {kind = "star"}
        cursor = cursor + (double and 2 or 1)
      end
    elseif char == "?" then
      tokens[#tokens + 1] = {kind = "question"}
      cursor = cursor + 1
    elseif char == "[" then
      local finish = class_end(pattern, cursor)
      if finish == nil then
        tokens[#tokens + 1] = {kind = "literal", value = char}
        cursor = cursor + 1
      else
        tokens[#tokens + 1] = parse_class(pattern, cursor, finish)
        cursor = finish + 1
      end
    else
      tokens[#tokens + 1] = {kind = "literal", value = char}
      cursor = cursor + 1
    end
  end
  return tokens
end

local function class_contains(token, char)
  if char == "" or char == "/" then return false end
  local matched = false
  local cursor = 1
  while cursor <= #token.characters do
    local first = token.characters[cursor]
    if cursor + 2 <= #token.characters and
        token.characters[cursor + 1] == "-" then
      local last = token.characters[cursor + 2]
      if first <= char and char <= last then matched = true end
      cursor = cursor + 3
    else
      if first == char then matched = true end
      cursor = cursor + 1
    end
  end
  return token.negated and not matched or (not token.negated and matched)
end

local function tokens_match(tokens, path)
  local memo = {}
  local function visit(token_index, path_index)
    local key = token_index .. ":" .. path_index
    if memo[key] ~= nil then return memo[key] end
    local token = tokens[token_index]
    if token == nil then
      local complete = path_index > #path
      memo[key] = complete
      return complete
    end

    local matched = false
    if token.kind == "literal" then
      matched = path:sub(path_index, path_index) == token.value and
        visit(token_index + 1, path_index + 1)
    elseif token.kind == "question" then
      local char = path:sub(path_index, path_index)
      matched = char ~= "" and char ~= "/" and
        visit(token_index + 1, path_index + 1)
    elseif token.kind == "class" then
      matched = class_contains(token, path:sub(path_index, path_index)) and
        visit(token_index + 1, path_index + 1)
    elseif token.kind == "star" then
      matched = visit(token_index + 1, path_index)
      local cursor = path_index
      while not matched and cursor <= #path and
          path:sub(cursor, cursor) ~= "/" do
        cursor = cursor + 1
        matched = visit(token_index + 1, cursor)
      end
    elseif token.kind == "double_star_directories" then
      matched = visit(token_index + 1, path_index)
      local cursor = path_index
      while not matched do
        local slash = path:find("/", cursor, true)
        if slash == nil then break end
        matched = visit(token_index + 1, slash + 1)
        cursor = slash + 1
      end
    elseif token.kind == "double_star_end" then
      matched = true
    end
    memo[key] = matched
    return matched
  end
  return visit(1, 1)
end

local function compile_pattern(body)
  local anchored = body:sub(1, math.max(0, #body - 1)):find("/", 1, true) ~= nil
  if body:sub(1, 1) == "/" then body = body:sub(2) end
  return {anchored = anchored, tokens = tokenize(body)}
end

local function pattern_matches(pattern, path)
  if pattern.anchored then return tokens_match(pattern.tokens, path) end
  if tokens_match(pattern.tokens, path) then return true end
  local cursor = 1
  while true do
    local slash = path:find("/", cursor, true)
    if slash == nil then return false end
    if tokens_match(pattern.tokens, path:sub(slash + 1)) then return true end
    cursor = slash + 1
  end
end

local function parse_pattern(raw)
  local body = strip_trailing_spaces(raw:gsub("\r$", ""))
  if body == "" or body:sub(1, 1) == "#" then return nil end
  if body:sub(1, 2) == "\\#" then body = body:sub(2) end

  local negated = false
  if body:sub(1, 1) == "!" then
    negated = true
    body = body:sub(2)
  elseif body:sub(1, 2) == "\\!" then
    body = body:sub(2)
  end
  if body == "" then return nil end

  local directory_only = false
  if body:sub(-1) == "/" and body:sub(-2) ~= "\\/" then
    directory_only = true
    body = body:sub(1, -2)
  end
  if body == "" then return nil end
  return {
    negated = negated,
    directory_only = directory_only,
    compiled = compile_pattern(body),
  }
end

local function parse_source(base_path, contents, precedence)
  local patterns = {}
  for line in (contents .. "\n"):gmatch("(.-)\n") do
    local pattern = parse_pattern(line)
    if pattern ~= nil then patterns[#patterns + 1] = pattern end
  end
  return {
    base_path = normalize(base_path),
    patterns = patterns,
    precedence = precedence,
  }
end

local function scoped_path(source, path)
  if source.base_path == "" then return path end
  local prefix = source.base_path .. "/"
  if path:sub(1, #prefix) ~= prefix then return nil end
  return path:sub(#prefix + 1)
end

local function source_verdict(source, path, is_directory)
  local scoped = scoped_path(source, path)
  if scoped == nil then return nil end
  local decision = nil
  for _, pattern in ipairs(source.patterns) do
    if (not pattern.directory_only or is_directory) and
        pattern_matches(pattern.compiled, scoped) then
      decision = not pattern.negated
    end
  end
  return decision
end

local function is_ignored(sources, path, is_directory)
  local ignored = false
  for _, source in ipairs(sources) do
    local verdict = source_verdict(source, path, is_directory)
    if verdict ~= nil then ignored = verdict end
  end
  return ignored
end

local function safe_read(read_text, path)
  local ok, contents = pcall(read_text, path)
  if not ok or type(contents) ~= "string" then return nil end
  return contents
end

local function depth(path)
  if path == "" then return 0 end
  local count = 1
  for _ in path:gmatch("/") do count = count + 1 end
  return count
end

function gitignore.sources(entries, read_text, start_path)
  local sources = {}
  local loaded = {}
  local function add(path, base_path, precedence)
    path = normalize(path)
    if loaded[path] then return end
    loaded[path] = true
    local contents = safe_read(read_text, path)
    if contents ~= nil then
      sources[#sources + 1] = parse_source(base_path, contents, precedence)
    end
  end

  add(".git/info/exclude", "", -1)
  add(".gitignore", "", 0)

  local start = normalize(start_path)
  local ancestor = ""
  for segment in start:gmatch("[^/]+") do
    ancestor = ancestor == "" and segment or ancestor .. "/" .. segment
    add(ancestor .. "/.gitignore", ancestor, depth(ancestor))
  end

  for _, entry in ipairs(entries or {}) do
    local path = normalize(entry.path or entry.name)
    if path == ".gitignore" then
      add(path, "", 0)
    elseif path:sub(-11) == "/.gitignore" and
        path:sub(1, 5) ~= ".git/" and
        not path:find("/.git/", 1, true) then
      local base_path = path:sub(1, -12)
      add(path, base_path, depth(base_path))
    end
  end

  table.sort(sources, function(left, right)
    if left.precedence ~= right.precedence then
      return left.precedence < right.precedence
    end
    return left.base_path < right.base_path
  end)
  return sources
end

local function is_git_metadata(path)
  return path == ".git" or path:sub(1, 5) == ".git/" or
    path:find("/.git/", 1, true) ~= nil or path:sub(-5) == "/.git"
end

function gitignore.is_visible(sources, raw_path, is_directory, include_ignored)
  local path = normalize(raw_path)
  if path == "" or is_git_metadata(path) then return false end
  if include_ignored then return true end

  local segments = {}
  for segment in path:gmatch("[^/]+") do segments[#segments + 1] = segment end
  local ancestor = ""
  local ancestor_count = is_directory and #segments or #segments - 1
  for index = 1, ancestor_count do
    ancestor = ancestor == "" and segments[index] or
      ancestor .. "/" .. segments[index]
    if is_ignored(sources, ancestor, true) then return false end
  end
  if is_directory then return true end
  return not is_ignored(sources, path, false)
end

function gitignore.glob_matches(pattern, path)
  return pattern_matches({anchored = true, tokens = tokenize(pattern)}, path)
end

function gitignore.relative_to(path, base_path)
  local normalized_path = normalize(path)
  local base = normalize(base_path)
  if base == "" then return normalized_path end
  local prefix = base .. "/"
  if normalized_path:sub(1, #prefix) ~= prefix then return nil end
  return normalized_path:sub(#prefix + 1)
end

return gitignore
